client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/RemoteBufferStreamReader.java

```diff
@@ -0,0 +1,122 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.celeborn.plugin.flink;
+
+import java.io.IOException;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.celeborn.common.network.protocol.ReadAddCredit;
+import org.apache.celeborn.plugin.flink.buffer.CreditListener;
+import org.apache.celeborn.plugin.flink.buffer.TransferBufferPool;
+import org.apache.celeborn.plugin.flink.network.MapShuffleClientImpl;
+import org.apache.celeborn.plugin.flink.network.RssBufferStream;
+import org.apache.celeborn.plugin.flink.protocol.ReadData;
+
+// This reader means that connection to a worker.
+public class RemoteBufferStreamReader extends CreditListener {
+  private static Logger logger = LoggerFactory.getLogger(RemoteBufferStreamReader.class);
+  private TransferBufferPool bufferPool;
+  private MapShuffleClientImpl client;
+  private String applicationId;
+  private int shuffleId;
+  private int mapId;
+  private int attemptId;
+  private int partitionId;
+  private int subPartitionIndexStart;
+  private int subPartitionIndexEnd;
+  private int currentSubPartitionIndex;
+  private boolean isOpen;
+  private Consumer<ByteBuf> dataListener;
+  private Consumer<Throwable> failureListener;
+  private RssBufferStream bufferStream;
+  private boolean closed = false;
+
+  public RemoteBufferStreamReader(
+      MapShuffleClientImpl client,
+      ShuffleResourceDescriptor shuffleDescriptor,
+      String applicationId,
+      int startSubIdx,
+      int endSubIdx,
+      TransferBufferPool bufferPool,
+      Consumer<ByteBuf> dataListener,
+      Consumer<Throwable> failureListener) {
+    this.client = client;
+    this.applicationId = applicationId;
+    this.shuffleId = shuffleDescriptor.getShuffleId();
+    this.mapId = shuffleDescriptor.getMapId();
+    this.attemptId = shuffleDescriptor.getAttemptId();
+    this.partitionId = shuffleDescriptor.getPartitionId();
+    this.bufferPool = bufferPool;
+    this.subPartitionIndexStart = startSubIdx;
+    this.subPartitionIndexEnd = endSubIdx;
+    this.dataListener = dataListener;
+    this.failureListener = failureListener;
+  }
+
+  public void open() throws IOException {
+    currentSubPartitionIndex = subPartitionIndexStart;
+    this.bufferStream =
+        client.readBufferedPartition(
+            applicationId, shuffleId, mapId, attemptId, partitionId, currentSubPartitionIndex);
```

Comment:
it's right that readBufferedPartition should read from subPartitionIndexEnd to subPartitionIndexEnd?

Reply:
`
int startSubIdx = gateDescriptor.getConsumedSubpartitionIndex();
        int endSubIdx = gateDescriptor.getConsumedSubpartitionIndex();
        checkState(endSubIdx >= startSubIdx);
        int numSubpartitionsPerChannel = endSubIdx - startSubIdx + 1;
`
This is the logic I copied from flink-remote-shuffle branch dev-flink-1.14.